src/runtime/pkg/govmm/qemu/qemu.go

```diff
@@ -2352,7 +2355,7 @@ type QMPSocket struct {
 
 // Valid returns true if the QMPSocket structure is valid and complete.
 func (qmp QMPSocket) Valid() bool {
-	if qmp.Type == "" || qmp.Name == "" {
+	if qmp.Type == "" || (qmp.Name == "" && qmp.FD == nil) {
```

Comment:
> The `false` condition should be `qmp.Type == "" || qmp.Name == "" || qmp.FD == nil`, correct? Otherwise, why do you even need to check whether `qmd.FD` is nil.

No. Either we pass the path that QEMU should bind the server socket to or we pass the fd of an already bound socket. In the latter case, the path already exists on the VFS and QEMU thus doesn't need it. It is thus valid to have one of Name of FD to be unset if the other one is set.

QEMU actually even prevents to pass both the path and the fd on the command line :

```
$ qemu-system-x86_64 -qmp unix:foo,fd=3,server,nowait
qemu-system-x86_64: -qmp unix:foo,fd=3,server,nowait: Exactly one of 'path', 'fd' or 'host' required
```

A stricter check is :

```
if qmp.Type == "" || (qmp.Name == "" && qmp.FD == nil) || (qmp.Name != "" && qmp.FD != nil) {
```

I'll rework this part to make it more obvious.

Reply:
FWIW, I think you can rewrite the stricter check as:
```
if qmp.Type == "" || (qmp.Name == "") != (qmp.FD == nil)
```